docs/src/main/asciidoc/config-reference.adoc

```diff
@@ -0,0 +1,778 @@
+This is a tradeoff as the more runtime properties are available, the less build time prework Quarkus can do. 
+The list of runtime properties is therefore lean.
+You can override these runtime properties with the following mechanisms (in decreasing priority) using:
+
+1. System properties
+2. Environment variables
+3. An environment file named `.env` placed in the current working directory
+4. A configuration file placed in `$PWD/config/application.properties`
+
+See <<configuration_sources>> for more details.
+
+== Generating configuration for your application
+
+It is also possible to generate an example `application.properties` with all known configuration properties, to make
+it easy to see what Quarkus configuration options are available. To do this, run:
+
+[source,bash]
+--
+./mvnw quarkus:generate-config
+--
+
+This will create a `src/main/resources/application.properties.example` file that contains all the config options
+exposed via the extensions you currently have installed. These options are commented out, and have their default value
+when applicable. For example this HTTP port config entry will appear as:
+
+
+[source,properties]
+--
+#
+# The HTTP port
+#
+#quarkus.http.port=8080
+--
+
+Rather than generating an example config file, you can also add these to you actual config file by setting the `-Dfile`
+parameter:
+
+[source,bash]
+--
+./mvnw quarkus:generate-config -Dfile=application.properties
+--
+
+If a config option is already present (commented or not) it will not be added, so it is safe to run this after
+adding an additional extension to see what additional options have been added.
+
+== Clearing properties
+
+Run time properties which are optional, and which have had values set at build time or which have a default value,
+may be explicitly cleared by assigning an empty string to the property.  Note that this will _only_ affect
+run time properties, and will _only_ work with properties whose values are not required.
+
+The property may be cleared by setting the corresponding `application.properties` property, setting the
+corresponding system property, or setting the corresponding environment variable.
+
+[[custom_configuration]]
+== Custom Configuration
+
+=== Custom configuration sources
+
+You can also introduce custom configuration sources in the standard MicroProfile Config manner.  To
+do this, you must provide a class which implements either `org.eclipse.microprofile.config.spi.ConfigSource`
+or `org.eclipse.microprofile.config.spi.ConfigSourceProvider`.  Create a
+https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html[service file] for the
+class and it will be detected and installed at application startup.
+
+=== Custom configuration converters
+
+You can also use your custom types as a configuration values. This can be done by implementing `org.eclipse.microprofile.config.spi.Converter<T>`
```

Comment:
```suggestion
You can also use your custom types for configuration values. This can be done by implementing `org.eclipse.microprofile.config.spi.Converter<T>`
```